### DISCUSSION TITLE
Auto-generate .htaccess files for specific folders?

### DISCUSSION THREAD

**Discussion by maschinenzeitmaschine** [11ty/eleventy#2543] [Q&A] [answered]
I have a portfolio website with several projects (each a page) and two global data files listing these projects:
- The first, 'projects.json' lists all pages which are linked and part of the website.
- The second, 'projects_unlisted.json', lists pages, I temporarily do not want to show and which are created, but not linked (so one can only view these with the direct url). I set this up in a way, so I can just move an entry from one into the other json, thereby in- or exluding the project from the main page.

Question: Is there any easy way to make eleventy copy a .htaccess file into the folders listed in the projects_unlisted.json, so these (and only these) can only be accessed when having a log-in and password?

The .htaccess file can be the same for all unlisted projects, something like this:
```
AuthName "open sesame"
AuthType Basic
AuthUserFile /path/.htpasswd
require valid-user
```

My projects.json and projects_unlisted.json look both the same, like this:
```
[
  {
    "shorttitle": "projectB",
    "thumb": "img/projectB.jpg",
    "url": "p_projectB/"
  },
  {
    "shorttitle": "projectE",
    "thumb": "img/projectE.jpg",
    "url": "p_projectE/"
  },
]
```

**Comment by pdehaan** [comment, marked as answer]
Easiest ways to add an .htaccess file are probably just using [passthrough file copy](https://www.11ty.dev/docs/copy/) or you could create an .htaccess.njk file and set a `permalink: unlisted/.htaccess` front matter key for that template.

**Reply by maschinenzeitmaschine** [reply]
Hi, thanks for helping! I'm not sure if I understand you correctly, but wouldn't that just copy/create a .htaccess file to some predetermined folder? Or can I use passthrough file copy dynamically from within a njk template?!? (i thougt, this only works in .eleventy.js.?)

Each of my pages are created as a folder with an index.html, and into each of these folders, or rather: only into each of the pages/folders listed in projects_unlisted.json, I would need to copy this .htaccess file… 

Or, maybe if that is clearer, what i would like to do, is run something like this at the end of building the site:
  {% for item in projects_unlisted %}
  copy .htacess file to {{ item.url }}
  {% endfor %}

**Reply by pdehaan** [reply]
Ah, perhaps I'm misunderstanding.
I thought you had something like:

- www/projects/one/index.html
- www/projects_unlisted/three/index.html
- www/projects_unlisted/nine/index.html

But it sounds like you want something more like:
- www/projects/one/index.html
- www/projects/three/
	- .htaccess
	- index.html
- www/projects/nine/
	- .htaccess
	- index.html

Yeah, the passthrough copy option would be in the .eleventy.js config file, so a bit tricky to do dynamically, unless you already know which projects should be unlisted (at which point it might be an easy loop and passthrough copy the same input file to _n_ different output folders).
Or possibly create some custom collection of unlisted projects, paginate over that, and dynamically set the `permalink` to each unlisted project output folder.

**Reply by pdehaan** [reply]
OK, I couldn't get the passthrough copy version to work since I guess it doesn't like me calling this multiple times w/ the same key:
```js
eleventyConfig.addPassthroughCopy({ ".htaccess": "p_projectA" });
eleventyConfig.addPassthroughCopy({ ".htaccess": "p_projectC" });
```
(it seems like it only gets copied into `p_projectC/.htaccess` since that was declared last)

But the pagination idea worked.

### .eleventy.js
```js
module.exports = function (eleventyConfig) {
  return {
    dir: {
      input: "src",
      output: "www",
    }
  };
};
```

### src/_data/projects_unlisted.json
```json
[
  {
    "shorttitle": "projectA",
    "thumb": "img/projectA.jpg",
    "url": "p_projectA/"
  },
  {
    "shorttitle": "projectC",
    "thumb": "img/projectC.jpg",
    "url": "p_projectC/"
  }
]
```

### src/projects/unlisted.njk
```njk
---
pagination:
  data: projects_unlisted
  size: 1
  alias: project
permalink: "{{ project.url }}"
---

<h1 data-unlisted>{{ project.shorttitle }}</h1>
```

### src/projects/unlisted_htaccess.njk
```njk
---
pagination:
  data: projects_unlisted
  size: 1
  alias: project
permalink: "{{ project.url }}.htaccess"
---

AuthName "open sesame"
AuthType Basic
AuthUserFile /.htpasswd
require valid-user
```

### OUTPUT
```sh
tree -aI node_modules
.
├── .eleventy.js
├── package-lock.json
├── package.json
├── src
│   ├── _data
│   │   ├── projects.json
│   │   └── projects_unlisted.json
│   └── projects
│       ├── listed.njk
│       ├── unlisted.njk
│       └── unlisted_htaccess.njk
└── www
    ├── p_projectA
    │   ├── .htaccess
    │   └── index.html
    ├── p_projectB
    │   └── index.html
    ├── p_projectC
    │   ├── .htaccess
    │   └── index.html
    └── p_projectE
        └── index.html

8 directories, 14 files
```

---

Another option might just be to have a single list/array of src/_data/projects.json and add an optional Boolean `unlisted` field:
```json
[
  {
    "shorttitle": "projectA",
    "thumb": "img/projectA.jpg",
    "url": "p_projectA/",
    "unlisted": true
  },
  {
    "shorttitle": "projectB",
    "thumb": "img/projectB.jpg",
    "url": "p_projectB/"
  },
  {
    "shorttitle": "projectC",
    "thumb": "img/projectC.jpg",
    "url": "p_projectC/",
    "unlisted": true
  },
  {
    "shorttitle": "projectE",
    "thumb": "img/projectE.jpg",
    "url": "p_projectE/"
  }
]
```

Although now you'd need a slight bit of extra work to filter the collection in order to dynamically write those .htaccess files in only the projects with an `unlisted: true` property (using 11ty pagination's [`before()` callback](https://www.11ty.dev/docs/pagination/#the-before-callback).

**Reply by pdehaan** [reply]
Actually, this was a bit easier than I expected (re: merging listed and unlisted projects into a single array):

### src/_data/projects.json

```json
[
  {
    "shorttitle": "projectA",
    "thumb": "img/projectA.jpg",
    "url": "p_projectA/",
    "unlisted": true
  },
  {
    "shorttitle": "projectB",
    "thumb": "img/projectB.jpg",
    "url": "p_projectB/"
  },
  {
    "shorttitle": "projectC",
    "thumb": "img/projectC.jpg",
    "url": "p_projectC/",
    "unlisted": true
  },
  {
    "shorttitle": "projectE",
    "thumb": "img/projectE.jpg",
    "url": "p_projectE/"
  }
]
```

### src/projects/pagination.njk
Write out BOTH listed and unlisted projects using the same template.
```njk
---
pagination:
  data: projects
  size: 1
  alias: project
permalink: "{{ project.url }}"
---

<h1>{{ project.shorttitle }}</h1>
```

### src/projects/htaccess.njk
Personally i generally prefer to move the JavaScript front matter into an *.11tydata.js file versus using `---js` front matter, but 🤷 .
```njk
---
# see htaccess.11tydata.js
---

AuthName "open sesame"
AuthType Basic
AuthUserFile /.htpasswd
require valid-user
```

### src/projects/htaccess.11tydata.js
And the front matter for the htaccess.njk file uses the `before()` callback to filter out ONLY the unlisted projects, so we can write the .htaccess files in those directories.
```js
module.exports = {
  pagination: {
    data: "projects",
    size: 1,
    alias: "project",
    before(data, cfg) {
      return data.filter(p => p.unlisted === true);
    }
  },
  permalink: "{{ project.url }}.htaccess"
};
```

### OUTPUT
```sh
> eleventy

[11ty] Writing www/p_projectA/index.html from ./src/projects/pagination.njk
[11ty] Writing www/p_projectB/index.html from ./src/projects/pagination.njk
[11ty] Writing www/p_projectC/index.html from ./src/projects/pagination.njk
[11ty] Writing www/p_projectE/index.html from ./src/projects/pagination.njk
[11ty] Writing www/p_projectA/.htaccess from ./src/projects/htaccess.njk
[11ty] Writing www/p_projectC/.htaccess from ./src/projects/htaccess.njk

[11ty] Wrote 6 files in 0.04 seconds (v1.0.2)
```

**Reply by maschinenzeitmaschine** [reply]
oh, wow! thank you so much for this custom-tailored, in-depth answer!!!
as i'd rather not change the setup with the two jsons (because that would cause other problems down the line), i went with your first route, but unfortunately i get an error:

```
[11ty] Problem writing Eleventy templates: (more in DEBUG output)
[11ty] > Output conflict: multiple input files are writing to `dist/p_projectA/index.html`. Use distinct `permalink` values to resolve this conflict.
  1. ./src/p_projectA.html
  2. ./src/unlisted.njk
```

i only see two things, which are different in your setup from mine, but i do not think either one is the problem... that said, they are:
- my project templates are not in src/projects/ but directly in src/
- as vscode does not format things nicely with .njk, but does so with .html, my project templates are named .html (but i did specify njk as sole template engine, so it should be all the same).

**Reply by pdehaan** [reply]
Ah, the old dreaded conflicting permalink issue.
Not sure I understand what is going wrong. Is ./src/p_projectA.html your input file?
I think I was dynamically generating the projects based on my mock JSON data from "src/_data/projects.json" and "src/_data/projects_unlisted.json". If you're manually creating those input files though, then you might not need my solution of the ./src/unlisted.njk file, and can just do the approach which copies just the .htaccess file into the unlisted folders:

```njk
---
pagination:
  data: projects_unlisted
  size: 1
  alias: project
permalink: "{{ project.url }}.htaccess"
---

AuthName "open sesame"
AuthType Basic
AuthUserFile /.htpasswd
require valid-user
```

Or this might be easier to debug if you have a public repo or simplified test case somewhere.

**Reply by maschinenzeitmaschine** [reply]
omg.
i have no idea how you manage to infer from my sparse error message, how i put this together, but you're right. i do have a separate template for each project and only use these datafiles for things like navigation. simply removing src/unlisted.njk did the trick. this works perfectly. again, thank you so very much!!!
